libs/confman/src/file_storage.cpp

```diff
@@ -0,0 +1,141 @@
+  if (const char* value = std::getenv(env_variable.c_str()); value != nullptr) {
+    if (int const value_i = std::stoi(value); value_i >= 0) {
+      return value_i;
+    }
+  }
+  return default_value;
+}
+
+/// \brief Fetches the value of the environment variable "TFC_CONFMAN_MIN_RETENTION_DAYS", default value is 30 days.
+/// \return `std::chrono::days` set in the environment variable
+/// \example get_minimum_retention_days();
+/// If the environment variable "TFC_CONFMAN_MIN_RETENTION_DAYS" is set to "15", the function returns std::chrono::days(15).
+/// Otherwise, the function returns std::chrono::days(30).
+static auto get_minimum_retention_days() -> std::chrono::days {
+  return std::chrono::days(get_environment_variable("TFC_CONFMAN_MIN_RETENTION_DAYS", 30));
+}
+
+/// \brief Fetches the value of the environment variable "TFC_CONFMAN_MIN_RETENTION_COUNT", default value is 4.
+/// \return minimum retention count set in the environment variable
+/// \example get_minimum_retention_count();
+/// If the environment variable "TFC_CONFMAN_MIN_RETENTION_COUNT" is set to "6", the function returns 6.
+/// Otherwise, the function returns 4.
+static auto get_minimum_retention_count() -> int {
+  return get_environment_variable("TFC_CONFMAN_MIN_RETENTION_COUNT", 4);
+}
+
+/// \brief Delete old files from the specified directory based on retention count and days. If the number
+/// of files exceed the retention count AND if the file's last modification date surpasses the retention time,
+/// \param directory path to directory containing files to possibly delete.
+static auto delete_old_files(std::filesystem::path const& directory) -> void {
+  int const retention_count = get_minimum_retention_count();
+  std::chrono::days const retention_time = get_minimum_retention_days();
+
+  // +2 for the file itself and the swap file
+  int total_file_retention_count = retention_count + 2;
+
+  if (int64_t const total_file_count =
+          std::distance(std::filesystem::directory_iterator(directory), std::filesystem::directory_iterator{});
+      total_file_count <= static_cast<int64_t>(total_file_retention_count)) {
+    return;
+  }
+
+  std::map<std::filesystem::file_time_type, std::filesystem::path> file_times;
+
+  for (const std::filesystem::directory_entry& entry : std::filesystem::directory_iterator(directory)) {
+    file_times.try_emplace(entry.last_write_time(), entry.path());
+  }
+
+  std::filesystem::file_time_type const current_time = std::filesystem::file_time_type::clock::now();
+
+  uint64_t count = 0;
+  for (auto it = file_times.begin();
+       it != file_times.end() && count <= (file_times.size() - static_cast<uint64_t>(total_file_retention_count));
+       ++it, ++count) {
+    auto time_since_last_modified = current_time - it->first;
+    if (time_since_last_modified > retention_time) {
+      std::filesystem::remove(it->second);
+    }
+  }
+}
```

Comment:
Please split this function up so you can unit test the user logic.
If the functions have side effects you make testing harder.

You can write "pure" functions and then string them together in a top level function.

Reply:
split the function and added more extensive testing, please see changes.